Isolated nonvolatile-storage capsule

atar13 opened the pull request:
### Pull Request Overview

This pull request adds isolation between userland apps to the nonvolatile_storage_driver
capsule. Previously, apps that used this capsule shared one large chunk or userspace
nonvolatile storage (with isolation between a section of kernel storage). This meant that
any application could read/write to any other applications nonvolatile
storage (which could potentially be sensitive).

With these changes, each app that uses nonvolatile storage will be
assigned a region of storage that is exclusive to them. From the
application's perspective, they will have a region of "logical storage"
which starts at address 0 and goes up to some size (which can be queried
from a syscall). This capsule maps these requests to the appropriate
locations in physical flash storage while checking that access requests
don't violate isolation guarantees.

See https://github.com/tock/tock/issues/3905 for more discussion and motivation for these changes.

#### Current Limitations:
- Each app has a fixed size of nonvolatile storage. This can be
  configured in the kernel when creating the nonvolatile storage
  capsule. Future implementations should consider giving each app more
  freedom over configuring the amount of nonvolatile memory they will use.
  This could be via a new system call (ex: reserve_storage(bytes: u32))
  or by a variable at flash time (ex: Makefile variable in libtock-c).
- Nonvolatile storage is only reserved when an app uses the capsule. On
  the first time that an app makes a syscall to this capsule, a fixed
  size region of storage is allocated for them. This means that if an app
  never uses this capsule, it will never have any storage allocated to
  it. Also, if too many apps call this capsule there might not be
  sufficient storage for every app. Initially, I wanted to do this
  reservation at capsule-creation time but I couldn't figure out how to
  populate grant memory at this time when the processes aren't loaded.
  I'm more than happy to adopt any ideas/suggestions on this front.
- Not the case after https://github.com/tock/tock/pull/4109/commits/e4bc0dbdb09c6bf56c5cd8219df72fe716a99c75 
  ~~No guarantees on ownership of nonvolatile storage through app or kernel re-flashing.
  The capsule doesn't store any persistent ownership data. When the
  kernel starts up again, it will hand out nonvolatile storage to any
  app that tries to use the capsule and will not respect any previous
  ownership of that region of storage.~~


### Testing Strategy

This pull request was tested with the [libtock-c `nonvolatile-storage` test ](https://github.com/tock/libtock-c/blob/master/examples/tests/nonvolatile_storage/main.c). I flashed multiple instances to ensure that each app wouldn't conflict with each other's isolated region of nonvolatile storage.


### TODO or Help Wanted

This pull request still needs feedback on when the best time to allocate each app's nonvolatile storage. As mentioned, currently it's done upon an app's first syscall to this capsule. However, this could be revised and any feedback would be appreciated.


### Documentation Updated

Will be done in a later libtock-c PR. ~~- [ ] Need to update libtock-c interface to explain how addresses are now "logical" (each app starts at 0) instead of absolute for all apps userspace.~~

### Formatting

- [x] Ran `make prepush`.

alistair23 commented:
> * No guarantees on ownership of nonvolatile storage through app or kernel re-flashing.
>       The capsule doesn't store any persistent ownership data. When the
>       kernel starts up again, it will hand out nonvolatile storage to any
>       app that tries to use the capsule and will not respect any previous
>       ownership of that region of storage.

Doesn't that defeat the point of long term storage?

bradjc commented:
This is very much needed. But we do need to use an app's `ShortId` to track its nonvolatile storage region so the app can keep its storage area across reboots.

atar13 commented:
This was a bit of an oversight on my end in the rush to get basic isolation up and running... I definitely agree that we need persistent ownership of these regions.

I think I'll adopt an approach similar to the one that @alistair23 mentioned [here](https://github.com/tock/tock/issues/3905#issuecomment-2078345425).

My one concern with that approach is the question of when nonvolatile storage will be freed. In that approach, it's mentioned that applications will be able to free their region of storage. I assume this will be through a syscall. However, I don't see a time during an applications life where it would want to free its nonvolatile storage. Most applications will reserve a bit of storage and get into a main loop where they never free it. This might not be a big deal just due to how abundant nonvolatile storage is. However, after a while, you might reflash your board with a bunch of different apps (ex: during development). Your nonvolatile storage will be filled with regions that belonged to apps that had previously been flashed. These apps would be hanging onto storage that they will never use again.

One idea I have to get around this is to provide a tockloader command to manage nonvolatile storage regions. This would allow the user to go in manually and decide which app's nonvoltiale storage they want to keep or purge.

Maybe this isn't as big of an issue as I think it is, but I'd love to get some more thoughts to figure out what design makes sense.

bradjc commented:
> My one concern with that approach is the question of when nonvolatile storage will be freed.

My opinion is we don't need to worry about that for this PR.

atar13 commented:
I'm in the midst of adding the headers to each app's region of storage to identify each region by the app that owns it. Things have been taking longer than anticipated as I've been getting up to speed with the Tock cell types and working with async reads/writes in the capsules. However, I've been making a lot of progress recently and hope to get the changes out soon.

I wanted to pose a question and get some input from others on how to handle a certain scenario.

At the start of each app's storage region, I'm adding two values to the header: the owning app's `ShortID` (u32) and the length of the app's region (usize). When the capsule is initialized, it will traverse these storage regions and store the location/length of each app's region to that app's grant. During this traversal stage, the capsule will not assume the lengths of each region are uniform, and it will calculate the starting address of the next region based on the length of the current region (which was just read from the header in persistent storage). 

This all works fine if we assume that nonvolatile storage has been populated with these regions and their correctly formatted headers. However, the first time the capsule is ever used the nonvolatile storage will be full of unknown, garbage data. The capsule has no knowledge of that fact and will try to interpret the first few bytes as a region header. This would mean that it would try to jump to the next region based on whatever value was sitting in storage where the region length value should have been.

One solution I had to this problem is to write a "magic" constant to the first few bytes of the entire userspace region of storage. When the capsule starts up, it could check for this value at a known location (start of userspace storage). If the value in storage matches the expected "magic" constant, it knows that the storage has been initialized previously, and we can trust the start of each region to be header information. If the value in storage does not match, we write the correct value and start allocating storage from the start of the userspace region. One potential issue is that the uninitialized data might happen to collide with the "magic" constant. However, we can reduce the chances of this happening by making the constant take up a large enough number of bits where it is extremely unlikely to collide. I'm not sure if this is the best solution, and it feels a bit hacky, so I'd love to get some thoughts and alternative approaches.

bradjc commented:
I think a `u32` magic value is the way to go. That's better than requiring an explicit init operation.

atar13 commented:
Following up on these changes with an update version that adds the persistence the original version was missing. Here is a comment from the top of the capsule that gives a high level overview of how the persistent isolated regions work:

Nonvolatile memory is reserved for each app when they explicitly call an initialization syscall. Note that only verified apps can reserve regions since this capsule uses the unique and persistent ShortID to identify the app across reboots. See this page in the Tock book for how to sign apps: <https://book.tockos.org/course/usb-security-key/key-hotp-access#signing-apps>
                                                                                          
Here is the sequence of events that happen when this initialization syscall is invoked:
 1. The capsule starts traversing a "linked-list" of app regions to find any existing regions that might already exist in storage for the app making the syscall. This traversal is possible due to headers that exist right before the start of each app's region. These headers describe the app that owns the region and size of the region. During traversal of these regions, the capsule uses the length to determine where the next region header starts.
 2. If the capsule finds a region header with a ShortID that matches the app that originally made the syscall, it saves the location and length of the region to the app's grant memory where it can be retrieved later during reads/writes.
 3. If the capsule doesn't find the region it will keep traversing until it reaches a termination point. The condition for terminating the traversal is when the capsule reads a header that has a ShortID value of 0. The value of 0 was chosen since the fixed variant of ShortIDs cannot take the value of 0. At this point, the capsule knows that app doesn't have a region assigned to it and therefore will assign it a new region at the location where the traveral ended. The length of the region is a fixed size that is configurable at compile time. Future implementations might want to focus on providing a way for apps to specify how much storage space they will need. Once a new region header is written, the capsule will write zeroes to the position of the next header to signify the new end of the "linked-list" of regions.
 4. Once an app is known to have a valid region (either by discovering it during traversal or allocating a new one), initialization completes and the app receives an upcall. Now it can go ahead and start reading/writing only within its isolated region.

Here is a storage layout diagram to illustrate where all these headers and regions live and how they fit into the existing kernel/userspace regions of storage:
```text
    ╒════════ ← Start of kernel region
    │
    │
    │
    ╞════════ ← Start of userspace region
    │ magic header (u32)
    ├──────── ← Start of App 1's region header
    │ App 1's ShortID (u32)  
    │ region 1 length (usize)
    ├──────── ← Start of App 1's Region          ═╗
    │                                             ║
    │
    │                                            region 1
    │                                            length
    │
    │                                             ║
    │                                            ═╝
    ├──────── ← Start of App 2's region header   
    │ App 2's ShortID (u32)                      
    │ region 2 length (usize)                    
    ├──────── ← Start of App 2's Region          ═╗
    │                                             ║
    │                                               
    │
    │                                            region 2
    │                                            length
    │
    │
    │                                             ║
    ...                                          ═╝
    ╘════════ ← End of storage
```

Feel free to ask any clarification questions if something doesn't make sense.

@alistair23 @bradjc

atar13 commented:
Also, I have been developing with the nrf52840dk and have modified the `boards/tutorials/nrf52840dk-thread-tutorial` board files to get app signing working according to the instructions on [this Tock book page](https://book.tockos.org/course/usb-security-key/key-hotp-access#verifying-credentials-in-the-kernel). These changes required adding a credential checker and changing the process loader. Should i go ahead and make these changes for other boards that have the nonvolatile_storage_driver capsule (ex: imix, stm32f3discovery, wm1110dev)?

ppannuto commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -38,6 +77,44 @@
+//!     │
+//!     │
+//!     │
+//!     ╞════════ ← Start of userspace region
+//!     │ magic header (u32)
+//!     ├──────── ← Start of App 1's region header
+//!     │ App 1's ShortID (u32)  
+//!     │ region 1 length (usize)
+//!     ├──────── ← Start of App 1's Region          ═╗
```
It maybe makes sense to include the `magic header` in each of the app's region headers? That has a few potential benefits:
 - Currently there is a hidden, second magic value (i.e. 0) which is used to demarcate the end of the list
    - And if `0` is missing for whatever reason, (e.g. a power failure when writing App 2's region header, but before it could write the "no App 3" header) this will do bad/buggy things.
 - The magic value could be used as a form of versioning for the region header
    - This may be more complex than helpful, but it could also be a kind of validity check for the header. i.e., `const MAGIC_HEADER_V1 = 0x....`, and you write to the region `magic_value = shortID ^ length ^ MAGIC_HEADER_V1` or similar

atar13 replied:
I'm in favor of adding a `magic_header` per region especially since it'll no longer require the terminating `0` magic value at the end (which I found kind of clunky).

Using it for versioning would be convenient as well since you would have versioning on a region by region basis. 

So based on the idea you mentioned, when the capsule reads the header it'll take `magic_value`, XOR it with `shortID` and `length`, and then check if it matches any of the supported `MAGIC_HEADER` versions (eg: `MAGIC_HEADER_V1 or MAGIC_HEADER_V2`) ?

ppannuto commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -128,10 +280,90 @@ impl Default for App {
     }
 }
 
+// the following helper functions are used for converting to/from
+// u8 slices that are read/written to nonvolatile storage
+fn u8_slice_to_usize(bytes: &[u8]) -> usize {
+    let mut result: usize = 0;
+
+    for (i, &byte) in bytes.iter().enumerate() {
+        result |= (byte as usize) << (8 * i);
+    }
+
+    result
+}
```
I believe you can use `usize::from_le_bytes` and similar for the methods here

atar13 replied:
Changed in https://github.com/tock/tock/pull/4109/commits/d02e6325cc22b2eeb101dac1dbdfd6263c2636a2.

ppannuto reviewed (changes requested):
Overall this looks great!

ppannuto commented:
re tutorials: They should be modified to still work at least. They don't necessarily all need to do full credential checking, e.g. using the 'accept any shortId in the TBF header' checker is probably sufficient for simple examples, rather than asking every tutorial to figure out app signing.

alistair23 commented:
> 3\. If the capsule doesn't find the region it will keep traversing until it reaches a termination point. The condition for terminating the traversal is when the capsule reads a header that has a ShortID value of 0

This seems a bit strange. If the backing storage is flash the values will all be `0xFF` after an erase operation. So something has to set the values to `0x00`. But then when we want to use it we need to erase the region before we actually use it.

That seems not ideal as we end up zeroing the entire region an excess time

alistair23 commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -98,19 +187,76 @@ mod rw_allow {
     pub const COUNT: u8 = 1;
 }
 
+/// Magic constant value written to the start of the entire userspace
+/// nonvolatile storage region. If the first 4 bytes (size of u32) of
+/// the userpace region match this magic constant, then we know the
+/// nonvolatile storage has been initialized by this capsule.
+const MAGIC_HEADER: u32 = 0x2FA7B3;
+/// Length of the above magic header value.
+const MAGIC_HEADER_LEN: usize = core::mem::size_of::<u32>();
```
We might want to include a version in this as well

atar13 replied:
Will version this as well as incorporate the versioned magic header into each region's header as mentioned here https://github.com/tock/tock/pull/4109#discussion_r1728059716

alistair23 commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -320,6 +1044,34 @@ impl<'a> NonvolatileStorage<'a> {
+                                    self.current_user.set(NonvolatileUser::HeaderManager(
+                                        HeaderState::Write(action),
+                                    ));
+                                    self.driver.write(header_buffer, offset, active_len)
```
So this writes the header, which is constructed in ` fn write_region_header()`. I guess that ends up calling `NonvolatileToPages` `write()`, but where doe the erase happen?

The value was previously 0 and now we are trying to set it, but I can't figure out where the erase happens. What am I missing here?

bradjc replied:
`NonvolatileStorage` allows us to just write.

bradjc replied:
Why not just not re-allocate regions, at least for now?

Re-allocating implies deciding when an application is "done" with a region, and that is an entirely different question we don't have an answer for. I'm not saying we should never address that question, but it doesn't seem important in the short term.

atar13 replied:
Yeah right now there is no concept of an app being "done" or freeing a region. The only way regions are re-used is if the magic header value is corrupted (which probably happens if the entire flash region is erased with some external tool from a computer (ex: nrfjprog)).

So maybe it's ok to just hand out regions without wiping them?

bradjc replied:
> So maybe it's ok to just hand out regions without wiping them?

I think so. The magic value and id are the record of the storage ownership; if those are corrupted, well, what can you do? We have to store that information _somewhere_ and then need to expect it remains.

alistair23 replied:
> `NonvolatileStorage` allows us to just write.

But where does the erase happen? I don't see it in the code

ppannuto replied:
In whatever implements the interface.

The `NonvolatileStorage` API is generic across underlying memory technologies, i.e. FRAM or Flash can both implement `NonvolatileStorage`. If the layer beneath is a flash layer that requires something like erase-before-write, then the underlying layer is responsible for doing the erase before doing the write.

alistair23 replied:
The interface doesn't either, it's a bug in `NonvolatileToPages` and the nRF flash implementation. I'll open an issue

alistair23 commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -202,9 +450,485 @@ impl<'a> NonvolatileStorage<'a> {
+            debug!("[NONVOLATILE_STORAGE_DRIVER]: Writing magic header");
+        }
+
+        let magic_header_slice = u32_to_u8_slice(MAGIC_HEADER);
+        self.header_buffer.map_or(Err(ErrorCode::NOMEM), |buf| {
+            // copy magic value to static buffer
+            for (i, c) in buf[0..magic_header_slice.len()].iter_mut().enumerate() {
+                *c = magic_header_slice[i];
+            }
+            Ok(())
+        })?;
+
+        self.enqueue_command(
+            NonvolatileCommand::HeaderWrite(HeaderWriteAction::WritingMagicHeader),
+            self.userspace_start_address,
+            magic_header_slice.len(),
+            None,
+        )
```
Even writing the magic header doesn't erase the flash.

Are you testing this against non-flash storage? We should probably make that clear then

atar13 replied:
> Even writing the magic header doesn't erase the flash.

Should the entire userspace flash area be erased when writing the valid magic header value or is it enough to just erase flash right before we hand it off to apps?

> Are you testing this against non-flash storage? We should probably make that clear then

No, I've only tested with flash storage. Should I clarify this at the top of the file?

bradjc replied:
This isn't flash specific because we are using the NonvolatileStorage HIL. This would work just fine with FRAM, for example.

alistair23 replied:
> Should the entire userspace flash area be erased when writing the valid magic header value or is it enough to just erase flash right before we hand it off to apps?

I don't think you should call erase. But it might be worth clearing the region before passing it to an app. You don't know what previous data was stored there and if we should expose it to an application

> No, I've only tested with flash storage. Should I clarify this at the top of the file?

Dose the nRF allow writes without erase? Is the flash already erased? `NonvolatileToPages` seems to never call erase, so this probably won't work on most flash devices.

That's not really something for this PR to fix, but maybe worth documenting somewhere

atar13 replied:
> I don't think you should call erase. But it might be worth clearing the region before passing it to an app. You don't know what previous data was stored there and if we should expose it to an application

Clearing the region makes sense and wouldn't need much code change. I already zero out the region header of the terminating region of the list, so I would just need to zero out the entire region before handing it to an app.

However, what do you think about pushing this off to a different PR that focuses more on freeing regions and reallocating them to different apps (as per the discussion here https://github.com/tock/tock/pull/4109#discussion_r1729477724)?

> Dose the nRF allow writes without erase? Is the flash already erased? NonvolatileToPages seems to never call erase, so this probably won't work on most flash devices.
> That's not really something for this PR to fix, but maybe worth documenting somewhere

I'm not too familiar with the inner workings of the nRF flash chip, but the `nrf52` `nvmc.rs` file seems to do some erasing before writing:

https://github.com/tock/tock/blob/8cef057ca61d941351912ae0ade63319a70bf913/chips/nrf52/src/nvmc.rs#L313-L319

I'm pretty sure this gets used since that same `nvmc` gets passed into the nonvolatile storage driver in the `main.rs` here: 

https://github.com/tock/tock/blob/8cef057ca61d941351912ae0ade63319a70bf913/boards/tutorials/nrf52840dk-thread-tutorial/src/main.rs#L168-L180

alistair23 replied:
> However, what do you think about pushing this off to a different PR that focuses more on freeing regions and reallocating them to different apps (as per the discussion here [#4109 (comment)](https://github.com/tock/tock/pull/4109#discussion_r1729477724))?

I think freeing regions is one thing. And can be delayed until later. When a region is freed it will also need to be cleared.

But I also think it's worth erasing a region before we pass it to userspace. We don't know what was previously there and it's possible it's something that we don't wan to expose.

alistair23 replied:
I opened https://github.com/tock/tock/issues/4152 about the lack of erase in `NonvolatileToPages`. It's a separate issue to this PR, but note that this implementation won't work on all Tock boards

atar13 replied:
> But I also think it's worth erasing a region before we pass it to userspace. We don't know what was previously there and it's possible it's something that we don't wan to expose.

Region content is now erased after a new region's header is written. https://github.com/tock/tock/pull/4109/commits/80714d383fc3a545b9c83d5c2f1a3d703e3330c1

atar13 commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -556,9 +1343,17 @@ impl SyscallDriver for NonvolatileStorage<'_> {
+                let res = self.apps.enter(processid, |app, _kernel_data| app.region);
+
+                // handle case where we fail to enter grant
+                res.map_or(CommandReturn::failure(ErrorCode::FAIL), |region| {
```
Changed in https://github.com/tock/tock/pull/4109/commits/294b9eb1aa557ea3c0213872ce3b85ae1e72d717!

atar13 commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -202,9 +450,485 @@ impl<'a> NonvolatileStorage<'a> {
+    fn allocate_app_region(&self, processid: ProcessId) -> Result<(), ErrorCode> {
+        // can't allocate a region if we haven't previously traversed existing regions
+        // and found where they stop
+        let Some(new_header_addr) = self.next_unallocated_region_header_address.get() else {
+            return Err(ErrorCode::FAIL);
+        };
```
The `.get()` method on OptionalCell returns an Option so there isn't really a meaningful error that gets bubbled up. Also I just learned that the `?` operator only works on Results and not Options.

The idea here is to make sure that we've previously traversed the regions in storage before allocating any new ones. So if we get to this function and we haven't finished traversing yet, something has gone wrong.

bradjc replied:
`?` works on both but the return type for the function has to match. I think you want https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or

atar13 replied:
`ok_or` is what I needed! Looks a lot cleaner now https://github.com/tock/tock/pull/4109/commits/daac50998a795667716fc5ed78c292c04d05feee

bradjc replied:
Yeah Rust has pretty good support for converting these types and handling errors, but unless you already know that it's possible (or that it _should_ be possible) it is really hard to discover that these APIs exist.

bradjc commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -202,9 +450,485 @@ impl<'a> NonvolatileStorage<'a> {
+                }
+            }
+            HeaderReadAction::ReadingRegionHeader(region_header_address) => {
+                // copy first few bytes from static buffer to local slice
+                let mut header_slice = [0; REGION_HEADER_LEN];
+                self.header_buffer.map_or(Err(ErrorCode::NOMEM), |buffer| {
+                    for (i, c) in buffer[0..header_slice.len()].iter().enumerate() {
+                        header_slice[i] = *c;
+                    }
+                    Ok(())
+                })?;
```
This shouldn't be necessary.

Also if it is helpful this file https://github.com/tock/tock/blob/master/libraries/tock-tbf/src/types.rs does a lot of converting buffers to various types.

atar13 replied:
I was able to replace most of the type conversions with bulitin rust functions (ex: `u32::from_le_bytes`).

However, these functions require a slice with a constant length. The buffer I get back from the closure has a type of `&mut [u8]` instead of something like `&mut [u8; 4]`. So right now I'm constructing a constant length buffer of the right size and copying the data over to it manually: 

https://github.com/tock/tock/blob/d02e6325cc22b2eeb101dac1dbdfd6263c2636a2/capsules/extra/src/nonvolatile_storage_driver.rs#L671-L673

jrvanwhy replied:
`&mut [u8]` implements [`TryInto<[u8; 4]>`](https://doc.rust-lang.org/stable/std/convert/trait.TryInto.html), so you should be able to write something like:

```rust
let owner = u32::from_le_bytes(owner_slice.try_into().unwrap());
```

atar13 replied:
Changed here: https://github.com/tock/tock/pull/4109/commits/a3f62a066ecd0a187efa2365584488f5ac3bb8a2. Nice to remove that extra copy!

bradjc reviewed (commented):
This looks awesome.

atar13 commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -202,9 +450,485 @@ impl<'a> NonvolatileStorage<'a> {
+        // Apps must have a fixed ShortID value. LocallyUnique apps
+        // are not allowed since we need a unique fixed ID to write to
+        // storage and use to identify apps.
+        let ShortId::Fixed(shortid) = processid.short_app_id() else {
+            return Err(ErrorCode::FAIL);
+        };
```
Ah I didn't even consider `NOSUPPORT`. I was considering `INVAL` for the cases missing the fixed shortid but `NOSUPPORT` makes more sense.

atar13 replied:
Changed in https://github.com/tock/tock/pull/4109/commits/caf20029630a503b43247763e47025b7e3e03ba7

bradjc commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -202,9 +450,479 @@ impl<'a> NonvolatileStorage<'a> {
+        // and found where they stop
+        let new_header_addr = self.next_unallocated_region_header_address.get().ok_or(ErrorCode::FAIL)?;
+
+        // Apps must have a fixed ShortID value. LocallyUnique apps
+        // are not allowed since we need a unique fixed ID to write to
+        // storage and use to identify apps.
+        let ShortId::Fixed(shortid) = processid.short_app_id() else {
+            return Err(ErrorCode::NOSUPPORT);
+        };
```
Oh I should have caught this earlier. Because this is nonvolatile storage, apps must have permission to use it. We need to use this (https://github.com/tock/tock/blob/master/kernel/src/storage_permissions.rs#L210) API, not `.short_app_id()`.

bradjc replied:
And then likewise we need to use the read permissions for reads, etc.

atar13 replied:
Should we store the `write_id` to storage in the region header instead of the `ShortID::Fixed` value? Or just use this API for checking permissions before allocating region and read/writes?

bradjc replied:
> Should we store the `write_id` to storage in the region header instead of the `ShortID::Fixed` value?

Yes.

(Of course it's the same, but the code will make more sense just using the storage permissions interface.)

atar13 replied:
Updated in this commit https://github.com/tock/tock/pull/4109/commits/002e08f4b5e52f08881759cf3c88c7578adde1c8.

If you could look over it, that would be great to make sure I'm using the API as intended.

However, with these changes my testing apps aren't able to allocate regions due to failing the new permission checks. I was looking over the TRD for Storage Permissions over here https://book.tockos.org/trd/trd-storage-permissions.html?highlight=permissions#7-specifying-permissions. 

So I need to implement a `PermissionsPolicy` trait in the board `main.rs` which chooses how to hand out app storage permissions?

atar13 commented:
With latest commit https://github.com/tock/tock/pull/4109/commits/3347436cd938443a3e4404753b8e4c61114fb94b, I've changed the mechanics of the magic header that was previously discussed as per this suggestion https://github.com/tock/tock/pull/4109#discussion_r1728059716.

Instead of having one single magic header value at the start of userpsace, this commit changes it so there is a magic value at the start of every region header. This magic value is set to the XOR of the region owner's ShortID, region length, and a constant to distinguish what version of the capsule the header was allocated from.

This has a few benefits:
- As this capsule is modified, it can support previous version of region headers.
- Previously, the capsule looked for a special ShortID value of 0 to determine if the region traversal was done. Now, it just checks if the magic value is invalid (not equal to an XOR between the region ShortID, length, and any of the supported version numbers). When an invalid header is reached, the traversal can be terminated and we can start allocating new regions there.

The diagram and step by step explanation of the header management at the top of `nonvolatiel_storage_driver.rs` have been updated.

atar13 commented:
My next TODO is to clear regions before handing them to apps based on the discussion here: https://github.com/tock/tock/pull/4109#discussion_r1731946150

bradjc commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -38,6 +79,45 @@
+//!     │
+//!     ╞════════ ← Start of userspace region
+//!     ├──────── ← Start of App 1's region header
+//!     │ App 1's magic value (usize) = HEADER_V1 ^ ShortID ^ length
+//!     │ App 1's ShortID (u32)  
+//!     │ region 1 length (usize)
```
With the disclaimer that this is pretty in the weeds given the context of this PR and how this capsule will be used, but:

This seems like an undesirable format because it requires speculatively checking for each version, and a chance of uncertainty. Or it requires each future version to guard against collisions with previous versions.

Imagine there was a version 2 and that the logic starts by assuming the region has a v2 header. What we wouldn't want is for a version 1 header to be decoded as a version 2. But it's not clear to me (given the general structure here) how a v2 header would be constructed such that it is impossible (particularly since we don't know anything about ShortId other than it is not 0) that these three fields and whatever data the app stored in its region could never match a valid v2 header.

That's why its common for the version to be stored separately and then there is no guessing what version the frame is.

This could be 

```
- 8 bits  - 24 bits                               -
| version | length                                |
| shortid                                         |
| xor                                             |
```

to keep it the same size.

atar13 replied:
That makes sense to have the version be stored separately instead of having to compute it and check it against all previous versions.

I just want to clarify the frame layout you showed. 

The first 32 bits are a concatenation of a u8 version and a 24 bit length. Then the next 32 bits are the shortid. And the last 32 bits are an xor between the last 2 32 bits.

ppannuto replied:
Yes, you are reading that diagram correctly

atar13 replied:
I went ahead and implemented this new header format.

However, I'm running into issues on determining if headers are valid. My first idea was to check if 
`xor == (version | length) ^ shortid`

I quickly realized this wouldn't work since I ran into an area of flash where all three fields of this frame were read in as 0. This satisfied the validity check since `0 == 0^0`.

Should I go back to matching the version number against all previous versions to show validity (in combination with my xor check)? Or is there another way I'm not thinking of checking validity with this format?

ppannuto replied:
Don't use version 0, start at 1 :)

[also be sure to shift `version << 24` in actual code if you aren't; just read this comment]

ppannuto replied:
If it's useful, you could split the current 8-bit version into 4-bit "magic" + 4-bit "version", such that there's something slightly more unique; not sure if that's necessary though.

atar13 replied:
I think the version number should be unique on its own without an extra magic number. I was using 1 as the V1 value as well.

I think I misinterpreted what Brad said [here](https://github.com/tock/tock/pull/4109#discussion_r1732888856) and thought that he didn't want me to check if the version number itself is one of the valid version numbers. But I think he meant that I shouldn't try to calculate the version from the result of an XOR operation (it's clearer just to read it in plainly).

So I think I'll check equality of the version I read as well as check that the `xor` field matches up.

atar13 replied:
The implementation of this change is up in this commit https://github.com/tock/tock/pull/4109/commits/327743ec1958a9558fccc00213aba778d336e801

bradjc commented on boards/tutorials/nrf52840dk-thread-tutorial/src/main.rs:
```diff
@@ -175,25 +205,34 @@ pub unsafe fn main() {
-        debug!("Error loading processes!");
-        debug!("{:?}", err);
-    });
+    let loader = static_init!(
```
We have a component for this, can it be used?

https://github.com/tock/tock/blob/master/boards/configurations/nrf52840dk/nrf52840dk-test-appid-sha256/src/main.rs#L290

atar13 replied:
Now using this component https://github.com/tock/tock/pull/4109/commits/6307c2ce06632fdc60228428f8e040098eae79f8

bradjc commented on boards/tutorials/nrf52840dk-thread-tutorial/src/main.rs:
```diff
@@ -132,6 +135,30 @@ pub unsafe fn main() {
+        .finalize(components::sha_software_256_component_static!());
+
+    // Create the credential checker.
+    let checking_policy = components::appid::checker_sha::AppCheckerSha256Component::new(sha)
+        .finalize(components::app_checker_sha256_component_static!());
+
```
We probably don't want to use this sha256 checker for now just to reduce complexity for running the tutorial. We can just use the null permit all checker.

atar13 replied:
Changed to use the null checker https://github.com/tock/tock/pull/4109/commits/6307c2ce06632fdc60228428f8e040098eae79f8

bradjc reviewed (commented):
This is looking very very cool. Much needed functionality.

I think this needs to be rebased so we can set the storage permissions policy on how we assign storage permissions to apps. A rebase will probably add the various policies and checkers we want as well.

atar13 commented:
> This is looking very very cool. Much needed functionality.
> 
> I think this needs to be rebased so we can set the storage permissions policy on how we assign storage permissions to apps. A rebase will probably add the various policies and checkers we want as well.

Just rebased and updated the thread tutorial to use the null checker, the name appid assigner, and to give each app individual storage permissions.

bradjc commented on boards/tutorials/nrf52840dk-thread-tutorial/src/main.rs:
```diff
@@ -203,28 +210,56 @@ pub unsafe fn main() {
+        NUM_PROCS
+    ));
+
+    checker.set_client(loader);
+
+    loader.register();
+    loader.start();
```
This all happens in the component and can be removed.

atar13 replied:
Updated here https://github.com/tock/tock/pull/4109/commits/91db6132e4f697da67d6cd3a310c45e676292f86

alistair23 commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -202,7 +464,427 @@ impl<'a> NonvolatileStorage<'a> {
+               region.offset 
+            );
+        }
+        self.region_erase_buffer.map_or(Err(ErrorCode::NOMEM), |buffer| {
+            // clear the erase buffer in case there was any nonzero value there
+            for c in buffer.iter_mut() {
+                *c = 0;
+            }
+            Ok(())
+        })?;
```
In future we should add a `clear()` function to the storage HIL so we can just call that, but this is ok for now

atar13 commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -54,22 +129,30 @@
 use capsules_core::driver;
+
+/// enable/disable debug prints
+const DEBUG: bool = false;
```
Removed all debugs https://github.com/tock/tock/pull/4109/commits/9b799dc6dc5e26732f9e50f8184335f4ad042362

atar13 commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -98,19 +183,167 @@ mod rw_allow {
+impl HeaderVersion {
+    fn value(&self) -> u8 {
+        match self {
+            HeaderVersion::V1 => HEADER_V1,
```
https://github.com/tock/tock/pull/4109/commits/02c0d74fe128b1410a75f99598bf55119ca009ec

atar13 commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -98,19 +183,167 @@ mod rw_allow {
 
+// Constants to distinguish what version of this capsule
+// a given header was created with.
+const HEADER_V1: u8 = 0x01;
```
https://github.com/tock/tock/pull/4109/commits/02c0d74fe128b1410a75f99598bf55119ca009ec

atar13 commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -202,7 +464,427 @@ impl<'a> NonvolatileStorage<'a> {
+                        || region.offset + region.length
+                            > self.userspace_start_address + self.userspace_length
+                    {
+                        return Err(ErrorCode::FAIL);
```
That makes more sense. Just changed https://github.com/tock/tock/pull/4109/commits/fb2b17e112716a64103bf002b5623bdb0b197923

bradjc commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -538,9 +1283,10 @@ impl SyscallDriver for NonvolatileStorage<'_> {
+    /// - `1`: Return the number of bytes available to each app.
     /// - `2`: Start a read from the nonvolatile storage.
     /// - `3`: Start a write to the nonvolatile_storage.
+    /// - `4`: Initialize an app's nonvolatile_storage.
```
How does this operation work? What happens if it is called more than once? What are the requirements for apps?

atar13 replied:
Init will either find an app's region that's already in storage or it will allocate a new one.

Apps must call this before any of the other syscalls (read/write/get_num_bytes) and wait for the init callback to fire.

There's also the requirements for setting up stuff in the `main.rs` like setting up `AppID` assignment and giving storage permissions but that's more of the responsibility of whoever is flashing the board.

bradjc commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -552,9 +1298,17 @@ impl SyscallDriver for NonvolatileStorage<'_> {
+                res.map_or(CommandReturn::failure(ErrorCode::NOMEM), |region| {
+                    // handle case where app's region is not assigned
+                    region.map_or(CommandReturn::failure(ErrorCode::FAIL), |region| {
+                        // TODO: Would break on 64-bit platforms
```
Why would this break on 64 bit platforms? Length can only be 24 bits right now, right?

atar13 replied:
Yeah, I can remove this. 

This was here since before I made any changes. I just left it but now it's inaccurate.

bradjc reviewed (commented):
This looks pretty close, but I do have a question about how init works, when an app must call init, what happens if init is called multiple times.

atar13 commented:
> This looks pretty close, but I do have a question about how init works, when an app must call init, what happens if init is called multiple times.

When init is called multiple times, the capsule will start traversing the iist of regions in storage by reading each of their headers. When the capsule finds a header with a shortid that matches one of the apps running on the board it will attempt to populate the app's grant with information about the location/length of the storage region. However, it performs a check to see if the storage region has already been populated in the grant. The app's region data is stored as an Option in the grant so it checks `app.region.is_none()`

https://github.com/atar13/tock/blob/05c7f928bc5154e82a1f8a6d9d64b8120c1f745c/capsules/extra/src/nonvolatile_storage_driver.rs#L689-L695
```rust
                        if app.has_requested_region && app.region.is_none() {
                            let version = header.version().ok_or(ErrorCode::FAIL)?;
                            app.region.replace(AppRegion {
                            ...
                            }
                         }

```

I could change the behavior so that the syscall will early exit if the app's region field is not `None` to avoid doing extra work?

bradjc commented:
> to populate the app's grant with information about the location/length of the storage region

Are apps supposed to call init? What if they don't?

Ideally we would just remove this to reduce complexity unless there is a specific use case. For example, if you think about writing a file from a linux process, we don't think about having to init the file system first before calling open().

ppannuto commented:
I wonder if the better analog (and maybe solution is simply name-bikeshedding) is that `init` is closer to `mount`? The `init` is somewhat heavyweight in that it will allocate space if-needed in flash / etc. Semantically, it's perhaps a bit nicer to separate the userspace concern of "have access to a filesystem" from "usage of filesystem". I suspect it keeps the implementation a bit simpler, as every storage operation [admittedly just read/write currently] does not need to have a possibly heavyweight/complex allocation check prepended.

Part of my thinking here is that an app being explicit about "mounting storage" opens the door in the future to an "unmount" equivalent which could be useful, especially in cases where it's costly in resources to keep pages/access "open" to storage.

bradjc commented:
In that case (where keeping something mounted is expensive) I would think the kernel wouldn't want to be beholden to the application saying it is ok to free up resources, it might just want to do that. If the overhead is in the grant region, then an app might want to be able to free that (I don't think this would actually work in practice due to the realities of the grant implementation, but makes sense in concept). This seems like what `open()` and `close()` do.

But I still think it is easier to just let apps use their storage and not have to worry about init/mount/open. This is just one storage API.

atar13 commented:
> > to populate the app's grant with information about the location/length of the storage region
> 
> 
> 
> Are apps supposed to call init? What if they don't?
> 
> 
> 
> Ideally we would just remove this to reduce complexity unless there is a specific use case. For example, if you think about writing a file from a linux process, we don't think about having to init the file system first before calling open().

Apps must call init to use nonvolatile storage. Reads and writes will fail otherwise.

I agree that this isn't a great abstraction. Initially, I wanted to do this when the capsule itself is initialized (maybe in `new()` or a capsule level `init()` function. However, I wasn't sure how to enter app grants at this point of execution. 

Essentially, I want to populate the grants with information about where any existing regions are located. So I go looking through storage for any region headers. Once I've read the region headers I can identify which app owned the region with its ShortID. However, I wasn't sure how to enter an apps grant with only its ShortID. As I understand, you need the ProcessID to enter a grant so I tried to find that. So my idea was to iterate through all the existing app grants `self.apps.iter()` and see which of their ShortIDs would match and use that element's ProcessID to enter the grant. However, I think this only worked once processes have been loaded and apps have already entered their grants. That's why I chose to make an explicit syscall before trying to store the region information in the grant. 

If that didn't make sense, I'm glad to clarify. I really would like to get rid of this init syscall but I'm not sure what a better alternative is. One idea I previously considered was to do the initialization sequence on the first syscall an app makes to this capsule. So that would mean that the first syscall would take longer because it would be doing this hidden initialization.

ppannuto commented:
> One idea I previously considered was to do the initialization sequence on the first syscall an app makes to this capsule. So that would mean that the first syscall would take longer because it would be doing this hidden initialization.

I believe that's Brad's suggestion. I don't think there are any promises on latency of operation for syscalls here. I think a longer first syscall is okay.

atar13 commented:
> I believe that's Brad's suggestion. I don't think there are any promises on latency of operation for syscalls here. I think a longer first syscall is okay.


Sounds good. I'll move with that approach. It'll avoid a big source of confusion from the app's perspective.

atar13 commented on capsules/extra/src/nonvolatile_storage_driver.rs:
```diff
@@ -552,9 +1298,16 @@ impl SyscallDriver for NonvolatileStorage<'_> {
+
+                // handle case where we fail to enter grant
+                res.map_or(CommandReturn::failure(ErrorCode::NOMEM), |region| {
+                    // handle case where app's region is not assigned
+                    region.map_or(CommandReturn::failure(ErrorCode::FAIL), |region| {
+                        CommandReturn::success_u32(region.length as u32)
+                    })
```
> I'm trying to think through this. The region size is fixed by the kernel. Should this just return that size, unconditionally? I guess this is related to the init question.

This should not return the fixed size unconditionally. Each region could potentially have a different size (if it was created with a previous kernel version). So, to maintain backwards compatibility we need to get the region size from what the region header reports.

> Because this is sync this cannot actually init. So we don't actually know if the app will get any space if it has not already been allocated space.

I was considering changing this to be an async function to allow it to complete the `init` operation. However, that's a breaking change to the syscall interface and requires changing existing user apps.

I like the idea of keeping it sync and having it fail immediately, however I feel like this will break some common scenarios.

An app might call this syscall as soon as it starts up to see how much space it has available (to allocate enough space for buffers). The syscall will fail since the app didn't previously call one of the other syscalls to trigger the initialization procedure. If the initialization Then the app will _have_ to know to call one of the other syscalls to kick off the init.

atar13 commented:
> I think it's ok to leave init but also automatically do it on first write. But we should clarify the semantics in the comment.

I think this makes sense, but should it only be on the first write? I think it makes sense to have it be on other syscalls such as reads and get_num_bytes. My idea is that anytime an app interacts with the capsule, the setup/init will be handled behind the scenes. The user won't have to worry about which specific syscall they call first.

bradjc commented:
> > I think it's ok to leave init but also automatically do it on first write. But we should clarify the semantics in the comment.
> 
> I think this makes sense, but should it only be on the first write? I think it makes sense to have it be on other syscalls such as reads and get_num_bytes. My idea is that anytime an app interacts with the capsule, the setup/init will be handled behind the scenes. The user won't have to worry about which specific syscall they call first.

Yes I think it has to be all calls.

atar13 commented:
> > I think this makes sense, but should it only be on the first write? I think it makes sense to have it be on other syscalls such as reads and get_num_bytes. My idea is that anytime an app interacts with the capsule, the setup/init will be handled behind the scenes. The user won't have to worry about which specific syscall they call first.
> 
> Yes I think it has to be all calls.

I'll go ahead and add it to all calls. However, I think this requires making the first syscall (get total available bytes) async with a callback (https://github.com/tock/tock/pull/4109#discussion_r1776046231)

bradjc commented:
FYI I'm digging into this and working on it. I'm realizing that the original code is quite old and we (I) have learned a lot about how to write capsules so there is a lot of cruft I want to remove.

atar13 commented:
> FYI I'm digging into this and working on it. I'm realizing that the original code is quite old and we (I) have learned a lot about how to write capsules so there is a lot of cruft I want to remove.

Sorry for not being too active on this recently, but I'd love to get the capsule up to the standard of the newer capsules. Is there a newer capsule I should check out as a reference?

bradjc commented:
My work is here: https://github.com/tock/tock/tree/isolated-nvm

The major changes are i) not requiring multiple static buffers, ii) using rust enums to contain state rather than N variables in the struct, iii) simplifying the logic so we can better reason about it (and fix bugs where buffers were `.take()`n and not replaced after an `if`).

atar13 commented:
Forgot to respond to this earlier. Those changes look great and solid improvements to the capsule. I especially like the consolidation of all the static buffers. Am I good to merge those into this PR?

One thing still lingering on my mind for this PR is figuring out how to add the initialization sequence to happen on the first syscall to this capsule.

We talked about it in a few comments earlier: https://github.com/tock/tock/pull/4109#discussion_r1768768613 and https://github.com/tock/tock/pull/4109#discussion_r1776046231

Adding the initialization to happen before the read and write syscalls is fine since if an error occurs during initialization, we can report it in the read/write callback. However, I can't figure out a good way of reporting an error for the get_num_bytes syscall. As we talked about, that one is synchronous so we can't just present an initialization error to the app via a callback. We definitely need to run the initialization during this syscall since apps will usually run this first to gauge how much storage they can read/write from. I just can't figure out how to handle errors without changing this syscall from sync to async.

atar13 commented:
After speaking with @bradjc , @ppannuto and @tyler-potyondy, we determined the best way forward is to move these changes to a new capsule (something like capsules_extra::isolated_nonvolatile_storage_driver). That way we can freely change the app interface (such as add a callback to the get_bytes syscall in case the initialization sequence fails).

atar13 commented:
@bradjc should I merge in your changes from https://github.com/tock/tock/tree/isolated-nvm ?

The commits are marked as WIP and I merged it locally and it wasn't compiling, so I assume you have more changes on your end.

bradjc commented:
I think what I have in https://github.com/tock/tock/tree/isolated-nvm is pretty close, but I ran out of time to get it to actually compile. You can keep working on it or I will try to get back to it soon.

The key changes are:
- I removed support for the kernel. I don't think that has ever been used, and that kind of made sense to support at the time I initially wrote that capsule but I don't think it belongs there now. We have better mechanisms to allow the kernel and userspace to share storage access.
- I simplified the code to use a single queue. That makes the state machine easier to reason about. The queue is entirely stored in app's grant regions now. But, we could add another effective element in the queue in the main struct in the future if we need support for additional operations.

atar13 commented:
Sounds good. Here's my next TODOs:

- Get your changes to compile and merge them in
- Move this code to a separate capsule (isolated_nonvolatile_storage_driver)
- Add change get_bytes syscall to error via callback

ppannuto commented:
With the 'split to parallel implementation', it might be easiest to cherry-pick code / commits and start a new branch / new PR here with the more clean-slate approach when it's ready.

atar13 commented:
> * Get your changes to compile and merge them in
> 
> * Move this code to a separate capsule (isolated_nonvolatile_storage_driver)

These two are added to my new branch https://github.com/atar13/tock/commits/isolated-nvm-driver/ in this commit https://github.com/atar13/tock/commit/44fcbf0b80434398ee744c6d5ff77127ec5f601a

> * Add change get_bytes syscall to error via callback

Going to get this added soon and make the PR

atar13 commented:
Made the new PR https://github.com/tock/tock/pull/4258 with the new capsule and the changes mentioned in the last few comments.

Going to close this PR in favor of that one.
